src/GrantRegistrar.sol

```diff
@@ -0,0 +1,203 @@
+  event WalletRegistered(address receiver, uint256 expiryGrandId);
+
+  /// @notice Emitted when the worldIdRouter is changed
+  /// @param worldIdRouter The new worldIdRouter instance
+  event WorldIdRouterUpdated(IWorldIDGroups worldIdRouter);
+
+  /// @notice Emitted when the groupId is changed
+  /// @param groupId The new groupId
+  event GroupIdUpdated(uint256 groupId);
+
+  /// @notice Emitted when the currentGrandId is changed
+  /// @param currentGrantId The new grantId
+  event CurrentGrantIdUpdated(uint256 currentGrantId);
+
+  /// @notice Emitted when the maxGrantId is changed
+  /// @param maxGrantId The new maxGrantId
+  event MaxGrantIdUpdated(uint256 maxGrantId);
+
+  ///////////////////////////////////////////////////////////////////////////////
+  ///                               CONSTRUCTOR                              ///
+  //////////////////////////////////////////////////////////////////////////////
+
+  /// @notice Deploys a WorldIDAirdrop instance
+  /// @param _worldIdRouter The WorldID router that will manage groups and verify proofs
+  /// @param _groupId The group ID of the World ID
+  /// @param _externalNullifierHash The nullifier hash that will be used to verify proofs
+  constructor(
+    IWorldIDGroups _worldIdRouter,
+    uint256 _groupId,
+    uint256 _externalNullifierHash,
+    uint256 _maxGrantId,
+    uint256 _currentGrantId
+  ) Ownable(msg.sender) {
+    if (address(_worldIdRouter) == address(0)) revert InvalidConfiguration();
+
+    groupId = _groupId;
+    maxGrantId = _maxGrantId;
+    worldIdRouter = _worldIdRouter;
+    currentGrantId = _currentGrantId;
+    externalNullifierHash = _externalNullifierHash;
+
+    emit GrantRegistrarInitialized(
+      worldIdRouter,
+      groupId,
+      externalNullifierHash,
+      maxGrantId,
+      currentGrantId
+    );
+  }
+
+  ///////////////////////////////////////////////////////////////////////////////
+  ///                               MAIN LOGIC                                ///
+  //////////////////////////////////////////////////////////////////////////////
+
+  /// @notice Registers a wallet to receive grants
+  /// @param receiver The address that will be able to claim grants
+  /// @param root The root of the Merkle tree (signup-sequencer or world-id-contracts provides this)
+  /// @param nullifierHash The nullifier for this proof, preventing double signaling
+  /// @param proof The zero knowledge proof that demonstrates the claimer has a verified World ID
+  /// @dev hashToField function docs are in lib/world-id-contracts/src/libraries/ByteHasher.sol
+  function verify(
+    address receiver,
+    uint256 root,
+    uint256 nullifierHash,
+    uint256[8] calldata proof
+  ) external payable {
+    if (nullifierHashes[nullifierHash] <= currentGrantId) revert InvalidNullifier();
+
+    worldIdRouter.verifyProof(
+      root,
+      groupId,
+      uint256(keccak256(abi.encodePacked(receiver, maxGrantId))) >> 8,
```

Comment:
hashToField?

Reply:
This is hashToField, but inlining the implementation since the source is not available in `world-id-contracts`. Same approach as with `RecurringGrantDrop` before.